### PR TITLE
Refactor determine prev and next step

### PR DESCRIPTION
**Pull request by frederike-ramin**
# Short Description
- This introduces the new logic how previous and next steps are determined.
- I did not change the eligibility logic though.

# Changes
- Introduce new method to determine prev and next steps
- Move the prev and next step logic of the lotse steps from the steps up to the step_chooser. This way we can use the logic in the lotse once we have more than one consecutive step as SteuerlotseStep
- Add `check_precondition? and `get_redirection_step` to the SteuerlotseStep
- Redirect in step_chooser if `get_redirection_step` of the step returns something

# Feedback
- Do you think we need a log/error message in case we return None as the prev/next step *and* the current step is not the first/last step (or is not in a defined list of accepted first/last steps)?
- What do you think of the way I handled the prev/next step logic for the lotse_step_chooser? Does that make sense to you? This way, we only have to set prev_step and next_step directly for the steps that direct to old multistepflow-steps.
- Do you see any missing tests?

### CODE REVIEW

File: webapp/tests/forms/flows/test_step_chooser.py [hunk 1]
```diff
@@ -101,6 +128,124 @@ def test_update_data_is_not_called_if_update_data_not_set(self):
...
+                                                                               {'precondition_met': True})
+        assert actual_next_step == expected_next_step
+
+    def test_if_next_steps_have_but_dont_meet_precondition_return_step_after(self, step_chooser_with_preconditions):
+        expected_next_step = MockFinalStep
+        actual_next_step = step_chooser_with_preconditions.determine_next_step(MockFormWithInputStep.name,
+                                                                               {'precondition_met': False})
+        assert actual_next_step == expected_next_step
+
+    def test_if_next_steps_have_but_dont_meet_precondition_request_correct_steps(self, step_chooser_with_preconditions):
```
Comment (JulianRoesner) [inline webapp/tests/forms/flows/test_step_chooser.py:231]: Maybe we can quickly talk together about that test?

**Review by JulianRoesner — APPROVED** [review]
> Do you think we need a log/error message in case we return None as the prev/next step and the current step is not the first/last step (or is not in a defined list of accepted first/last steps)?

That is a very good question and I am not so sure. I am wondering if we could swap it and just return None for these very specific cases and really raise an error in the other cases but maybe let's discuss this.

> What do you think of the way I handled the prev/next step logic for the lotse_step_chooser? Does that make sense to you? This way, we only have to set prev_step and next_step directly for the steps that direct to old multistepflow-steps.

To be honest it makes sense but it is a little tricky to understand. I think we should add some comments to understand what is happening. It took me quite some time to find my way back into what is happening there.

> Do you see any missing tests?
So far no. But maybe it would help a lot if we could quickly go through it together

File: webapp/app/forms/steps/lotse/personal_data.py [hunk 2]
```diff
@@ -19,16 +19,14 @@
 class LotseFormSteuerlotseStep(FormSteuerlotseStep):
```
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/forms/steps/lotse/personal_data.py:19]: The precondition is set to None in the SteuerlotseStep. And per default we don't need a precondition for all Lotse steps, do we?

File: webapp/tests/forms/flows/test_step_chooser.py (hunk 1 above)
Comment (frederike-ramin) [reply webapp/tests/forms/flows/test_step_chooser.py:235]: rename to `request_direct_successors`

**Comment by frederike-ramin** [comment]
Todos left on this PR:
- [ ] Go over the tests together
- [ ] Add a comment explaining the detemrine_x_step functions
- [ ] Return None only for specific steps and error by default

File: webapp/tests/forms/flows/test_step_chooser.py [hunk 3]
```diff
@@ -101,6 +132,124 @@ def test_update_data_is_not_called_if_update_data_not_set(self):
...
+             patch('tests.forms.mock_steuerlotse_steps.MockMiddleStep.check_precondition') as middle_step_check, \
+             patch('tests.forms.mock_steuerlotse_steps.MockFormWithInputStep.check_precondition') as input_step_check, \
+             patch('tests.forms.mock_steuerlotse_steps.MockFinalStep.check_precondition') as final_step_check:
+            step_chooser_without_preconditions.determine_prev_step(MockFormWithInputStep.name, {})
+            start_step_check.assert_not_called()
+            middle_step_check.assert_called_once()
+            input_step_check.assert_not_called()
+            final_step_check.assert_not_called()
+
+    def test_if_previous_steps_have_met_precondition_then_return_direct_predecessor(self, step_chooser_with_preconditions):
```
Comment (JulianRoesner) [inline webapp/tests/forms/flows/test_step_chooser.py:164]:
```suggestion
    def test_if_previous_step_have_met_precondition_then_return_direct_predecessor(self, step_chooser_with_preconditions):
```

**Pull request merged**